plugins/catalog-backend-module-gitlab/src/providers/GitlabDiscoveryEntityProvider.ts

```diff
@@ -207,10 +208,110 @@ export class GitlabDiscoveryEntityProvider implements EntityProvider {
       );
     }
 
+    const locations = await this.getEntities();
+
+    await this.connection.applyMutation({
+      type: 'full',
+      entities: locations.map(location => ({
+        locationKey: this.getProviderName(),
+        entity: locationSpecToLocationEntity({ location }),
+      })),
+    });
+
+    logger.info(`Processed ${locations.length} locations`);
+  }
+
+  /**
+   * Determine the location on GitLab to be ingested base on configured groups and filters.
+   *
+   * @returns A list of location to be ingested
+   */
+  private async getEntities() {
+    let res: Result = {
+      scanned: 0,
+      matches: [],
+    };
+
+    const groupToProcess = new Map<string, GitLabGroup>();
+
+    if (
+      this.config.groupPatterns !== undefined &&
+      this.config.groupPatterns.length > 0
+    ) {
+      const groups = paginated<GitLabGroup>(
+        options => this.gitLabClient.listGroups(options),
+        {
+          page: 1,
+          per_page: 50,
+        },
+      );
+
+      for (const pattern of this.config.groupPatterns) {
+        for await (const group of groups) {
```

Comment:
i don't think you want to do the _server_ retrieval many times if there are many patterns. This `for` loop should probably be the outer one, and then the inner one should do something like `if (this.config.groupPatterns.some(...` or whatnot